### PR TITLE
Warn when access keys are old

### PR DESCRIPTION
**Pull request by TimaJuma**
#### CHANGES

 - Added command to update Access Key expiry period (in days)
 - When 
+ `warn` if secretKey is about to expire (today, tomorrow or has expired)

> without Access Key Max Age
image

> Set Access Key Max Age

set-max-age

> When awsAccessKeyMaxAge is set
- Fresh Key
image

-Stale Key
image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "awsx",
   "description": "AWS CLI profile switcher with MFA support",
-  "version": "1.3.0",
+  "version": "1.3.1",
```
Comment (TimaJuma) [inline package.json:4]: not sure if it should be `1.3.1` or `1.4.0`

File: src/app.ts [hunk 2]
```diff
@@ -670,14 +684,15 @@ const awsx = (): void => {
     })
     .command({
       command:
-        'add-profile [profile] [access-key] [secret-key] [default-region] [output-format] [mfa-arn] [mfa-expiry]',
+        'add-profile [profile] [access-key] [secret-key] [secret-key-expiry] [default-region] [output-format] [mfa-arn] [mfa-expiry]',
```
Comment (TimaJuma) [inline src/app.ts:687]: For existing `profiles` added command to update profile with [secret-key-expiry] field

File: src/app.ts [hunk 3]
```diff
@@ -912,8 +912,8 @@ const awsx = (): void => {
       },
     })
     .command({
-      command: 'add-key-expiry [profile] [expiry-period]',
-      describe: "Add expiry to your profile's secret key",
+      command: 'add-key-max-age [profile] [max-age]',
+      describe: 'Add max age of profile access key in days',
```
Comment (rogersnick) [inline src/app.ts:916]: can we put an example of the command here to make it easy for the user to copy / paste?

File: src/command/check-secret-expiry.ts [hunk 4]
```diff
@@ -3,29 +3,47 @@ import chalk from 'chalk';
...
+        text: `${daysLeft === 1 ? '⚠ expires tomorrow' : '⚠ expires today'}`,
+        textColor: chalk.yellow,
+      };
+    } else if (daysLeft < 7) {
+      return { text: `⚠ expires in ${daysLeft} days`, textColor: chalk.yellow };
+    } else if (daysLeft < 30) {
+      return {
+        text: `✓ expires in ${keyAge} days`,
+        textColor: chalk.green,
+      };
```
Comment (rogersnick) [inline src/command/check-secret-expiry.ts:32]: we should change all of this text to be clearer that we are approaching key-max-age, and not that the key is actually expiring. Something like Key will reach maximum age in X days maybe? I think that we just don't want to tell the user their keys expired if it actually isn't

File: src/command/check-secret-expiry.ts [hunk 5]
```diff
@@ -54,20 +76,32 @@ const checkSecretKeyExpiry = async (
...
-    const expiryPeriod = profile.awsSecretAccessKeyExpiry || DEFAULT_SECRET_KEY_EXPIRY_DAYS;
-
-    const [expired, chalkColor] = getExpiryStatus(currentKey.CreateDate, expiryPeriod);
+    const keyMaxAge = profile.awsAccessKeyMaxAge;
+    const expiryStatus = getExpiryStatus(currentKey.CreateDate, keyMaxAge);
 
-    if (expired && chalkColor) {
+    if (keyMaxAge && expiryStatus?.text) {
+      console.log(
+        expiryStatus.textColor(`Your AccessKey of profile ${profile.profileName} has expiry period of ${keyMaxAge} days.
```
Comment (rogersnick) [inline src/command/check-secret-expiry.ts:84]: here as well, something like `Maximum Access Key age is set to ____. `

**Review by rogersnick — CHANGES_REQUESTED** [review]

File: src/app.ts [hunk 6]
```diff
@@ -890,6 +911,31 @@ const awsx = (): void => {
         }
       },
     })
+    .command({
+      command: 'add-key-max-age [profile] [max-age]',
+      describe: 'Add max age of profile access key in days',
```
Comment (iansu) [inline src/app.ts:916]:
```suggestion
      command: 'set-key-max-age [profile] [max-age]',
      describe: 'Set maximum age of AWS access key in days',
```

File: src/command/add-key-max-age.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      {
+        type: 'select',
+        name: 'profile',
+        message: 'Choose a profile',
+        choices: choices,
+      },
+      {
+        type: 'number',
+        name: 'period',
+        message: 'Enter access key maximum age in days',
```
Comment (iansu) [inline src/command/add-key-max-age.ts:29]:
```suggestion
        message: 'Access key maximum age in days (use 0 for no maximum age)',
```

File: src/command/check-secret-expiry.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    userName = userName
+      ? userName
+      : getUserName((await createStsClient(profile).getCallerIdentity({})).Arn);
+
+    const iamClient = createIAMClient(credentials, profile.awsDefaultRegion);
+    const keys = await iamClient.send(new ListAccessKeysCommand({ UserName: userName }));
+
+    const currentKey = keys.AccessKeyMetadata?.find(
+      (k) => k.AccessKeyId === profile.awsAccessKeyId
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:72]:
```suggestion
      (key) => key.AccessKeyId === profile.awsAccessKeyId
```

File: src/command/check-secret-expiry.ts (hunk 4 above)
Comment (iansu) [reply src/command/check-secret-expiry.ts:32]: I agree we should not be talking about expiration. It's also a bit hard to follow all the different paths in this code. Can you provide some examples of the different outputs? Either just a text version in a comment or a screenshot?

**Review by iansu — CHANGES_REQUESTED** [review]
Looks pretty good. Just a few display/wording changes and I think we're good.

File: src/command/check-secret-expiry.ts [hunk 9]
```diff
@@ -6,43 +6,38 @@ import { AWSCredentials, ProfileConfiguration } from '../mfa-login';
...
+
+  if (daysLeft < 0) {
+    return { text: `✘ maximum age is exceeded`, textColor: chalk.red };
+  } else if ([0, 1].includes(daysLeft)) {
+    return {
+      text: `${daysLeft === 1 ? '⚠ maximum age exceeds tomorrow' : '⚠ maximum age exceeds today'}`,
+      textColor: chalk.yellow,
+    };
+  } else if (daysLeft < 7) {
+    return { text: `⚠ maximum age exceeds in ${daysLeft} days`, textColor: chalk.yellow };
```
Comment (rogersnick) [inline src/command/check-secret-expiry.ts:23]: exceeds --> `will be reached`

**Review by rogersnick — COMMENTED** [review]
quick copy comment

**Review by rogersnick — COMMENTED** [review]
could use a couple screenshots to see how this looks to users

**Review by usernameseb — DISMISSED** [review]
Looks good to me, thanks for adding those screenshots

File: src/command/check-secret-expiry.ts [hunk 10]
```diff
@@ -13,14 +13,18 @@ const getKeyAgeStatus = (
   const daysLeft = keyMaxAge - keyAge;
 
   if (daysLeft < 0) {
-    return { text: `✘ maximum age is exceeded`, textColor: chalk.red };
+    return { text: `✘ maximum age has exceeded`, textColor: chalk.red };
```
Comment (rogersnick) [inline src/command/check-secret-expiry.ts:16]: dude. 
I told you this should say `has been exceeded` 
I explicitly left that as a comment.

**Review by rogersnick — CHANGES_REQUESTED** [review]

File: src/command/check-secret-expiry.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+          ? '⚠ maximum age will be reached tomorrow'
+          : '⚠ maximum age has been reached today'
+      }`,
+      textColor: chalk.yellow,
+    };
+  } else if (daysLeft < 7) {
+    return { text: `⚠ maximum age will be reached in ${daysLeft} days`, textColor: chalk.yellow };
+  } else if (daysLeft < 30) {
+    return {
+      text: `✓ maximum age exceeds in ${keyAge} days`,
```
Comment (rogersnick) [inline src/command/check-secret-expiry.ts:30]: same thing here

File: src/command/check-secret-expiry.ts (hunk 9 above)
Comment (rogersnick) [reply src/command/check-secret-expiry.ts]: Don't resolve things if you don't make changes or leave a comment please

File: src/command/check-secret-expiry.ts (hunk 10 above)
Comment (rogersnick) [reply src/command/check-secret-expiry.ts:16]: sorry I actually said "will be reached" ... either way -- you just resolved the comment, made no changes, and then re-requested me.

File: src/command/check-secret-expiry.ts (hunk 9 above)
Comment (TimaJuma) [reply src/command/check-secret-expiry.ts]: I am a bit confused.
image

I think your comment was referring to line 19 and 23 originally, and I changed the text accordingly. 
Thats why it was resolved.

File: src/command/check-secret-expiry.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+      textColor: chalk.green,
+    };
+  } else return;
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:33]: This is unnecessary and can be removed

File: src/command/check-secret-expiry.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+import { createIAMClient, createStsClient } from '../lib/create-aws-client';
+
+const getKeyAgeStatus = (
+  keyAge: number,
+  keyMaxAge: number
+): { text: string; textColor: chalk.Chalk } | undefined => {
+  const daysLeft = keyMaxAge - keyAge;
+
+  if (daysLeft < 0) {
+    return { text: `✘ maximum age has exceeded`, textColor: chalk.red };
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:16]: Here's how I think this message should read:

> 30 days: show nothing
< 30 days: `Access key is X days old (maximum age is Y days)` (info icon/blue text)
< 7 days: `Access key is X days old (maximum age is Y days)` (warning icon/yellow text)
< 0 days: `Access key is X days old (maximum age is Y days)` (error icon/red text)

File: src/command/check-secret-expiry.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+    const keyMaxAge = profile.awsAccessKeyMaxAge;
+    const keyAge = getKeyAgeInDays(currentKey.CreateDate.getTime());
+
+    if (keyMaxAge) {
+      const status = getKeyAgeStatus(keyAge, keyMaxAge);
+
+      if (!status) return;
+
+      console.log(
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:86]: I don't think we need this message at all, just the message from the `getKeyAgeStatus` function (which should probably be renamed to something like `getKeyAgeMessage`)

File: src/command/check-secret-expiry.ts [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+      return;
+    }
+
+    const keyMaxAge = profile.awsAccessKeyMaxAge;
+    const keyAge = getKeyAgeInDays(currentKey.CreateDate.getTime());
+
+    if (keyMaxAge) {
+      const status = getKeyAgeStatus(keyAge, keyMaxAge);
+
+      if (!status) return;
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:84]: Don't use conditionals without a body

File: src/command/check-secret-expiry.ts [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+AccessKey status : ${status.text}
+`)
+      );
+    } else {
+      const textColor = getTextColor(keyAge);
+
+      if (!textColor) return;
+
+      console.log(
+        textColor(`Your AccessKey of profile ${profile.profileName} is ${keyAge} days old.`)
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:97]: This should just say `Access key is X days old`

File: src/command/check-secret-expiry.ts [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+      );
+    }
+
+    console.log(
```
Comment (iansu) [inline src/command/check-secret-expiry.ts:101]: This should only be shown if the user hasn't customized the key max age. It should say `Change this setting with 'awsx set-key-max-age [profile] <days>' (use 0 days for no maximum age)`

**Review by iansu — CHANGES_REQUESTED** [review]

File: src/command/check-secret-expiry.ts [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    if (!currentKey || !currentKey.CreateDate) {
+      return;
+    }
+
+    const keyMaxAge = profile.awsAccessKeyMaxAge;
+    const keyAge = getKeyAgeInDays(currentKey.CreateDate.getTime());
+
+    if (keyMaxAge) {
+      printKeyMaxAgeMessage(keyAge, keyMaxAge);
+      console.log();
```
Comment (TimaJuma) [inline src/command/check-secret-expiry.ts:86]: empty line

File: package.json [hunk 19]
```diff
@@ -69,7 +69,7 @@
     "rimraf": "3.0.2",
     "rollup": "^2.60.2",
     "rollup-plugin-terser": "^7.0.2",
-    "rollup-plugin-uglify": "^6.0.4",
+    "rollup-plugin-uglify": "^5.0.2",
```
Comment (rogersnick) [inline package.json:72]: Any reason we're downgrading this and husky (above)?

Comment (TimaJuma) [reply package.json:72]: npm audit

Comment (rogersnick) [reply package.json:72]: I suppose brevity is the soul of wit

**Review by rogersnick — APPROVED** [review]

**Review by iansu — CHANGES_REQUESTED** [review]
This looks great. I've been trying it out locally and it seems to work well. One last thing though, you should update the README to include the new command to set the key age. Once that's done I think this is ready to merge.

File: README.md [hunk 20]
```diff
@@ -44,7 +44,7 @@ If you don't specify a profile name (or an assume role profile name, if applicab
 
 ### Adding a new profile
 
-#### `awsx add-profile [profile] [access-key] [secret-key] [default-region] [output-format] [mfa-arn] [mfa-expiry]`
+#### `awsx add-profile [profile] [access-key] [secret-key] [key-max-age] [default-region] [output-format] [mfa-arn] [mfa-expiry]`
```
Comment (TimaJuma) [inline README.md:47]: I will update wiki once ticket is approved

**Review by rogersnick — APPROVED** [review]

**Review by iansu — APPROVED** [review]
